plugin/path_role.go

```diff
@@ -444,68 +555,196 @@ func (role *gcpRole) updateRole(sys logical.SystemView, op logical.Operation, da
+			return err
+		}
+		if err := role.updateGceFields(data, op); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (role *gcpRole) validate(sys logical.SystemView) (warnings []string, err error) {
+	warnings = []string{}
+
+	switch role.RoleType {
+	case iamRoleType:
+		if warnings, err = role.validateForIAM(); err != nil {
+			return warnings, err
+		}
+	case gceRoleType:
+		if warnings, err = role.validateForGCE(); err != nil {
+			return warnings, err
+		}
+	case "":
+		return warnings, errors.New(errEmptyRoleType)
 	default:
-		return fmt.Errorf("role type '%s' is invalid", role.RoleType)
+		return warnings, fmt.Errorf("role type '%s' is invalid", role.RoleType)
 	}
 
 	if len(role.Policies) == 0 {
-		return errors.New("role must have at least one bound policy")
+		return warnings, errors.New("role must have at least one bound policy")
 	}
 
 	if role.ProjectId == "" {
-		return errors.New("role cannot have empty project_id")
+		return warnings, errors.New(errEmptyProjectId)
 	}
 
-	if role.MaxJwtExp > time.Hour {
-		return errors.New("max_jwt_exp cannot be more than one hour")
+	defaultLeaseTTL := sys.DefaultLeaseTTL()
+	if role.TTL > defaultLeaseTTL {
+		warnings = append(warnings, fmt.Sprintf(
+			"Given ttl of %d seconds greater than current mount/system default of %d seconds; ttl will be capped at login time",
+			role.TTL/time.Second, defaultLeaseTTL/time.Second))
 	}
 
+	defaultMaxTTL := sys.MaxLeaseTTL()
+	if role.MaxTTL > defaultMaxTTL {
+		warnings = append(warnings, fmt.Sprintf(
+			"Given max_ttl of %d seconds greater than current mount/system default of %d seconds; max_ttl will be capped at login time",
+			role.MaxTTL/time.Second, defaultMaxTTL/time.Second))
+	}
 	if role.MaxTTL < time.Duration(0) {
-		return errors.New("max_ttl cannot be negative")
+		return warnings, errors.New("max_ttl cannot be negative")
 	}
 	if role.MaxTTL != 0 && role.MaxTTL < role.TTL {
-		return errors.New("ttl should be shorter than max_ttl")
+		return warnings, errors.New("ttl should be shorter than max_ttl")
 	}
 
 	if role.Period > sys.MaxLeaseTTL() {
-		return fmt.Errorf("'period' of '%s' is greater than the backend's maximum lease TTL of '%s'", role.Period.String(), sys.MaxLeaseTTL().String())
+		return warnings, fmt.Errorf("'period' of '%s' is greater than the backend's maximum lease TTL of '%s'", role.Period.String(), sys.MaxLeaseTTL().String())
 	}
 
-	return nil
+	return warnings, nil
 }
 
 // updateIamFields updates IAM-only fields for a role.
 func (role *gcpRole) updateIamFields(data *framework.FieldData, op logical.Operation) error {
+	allowGCEInference, ok := data.GetOk("allow_gce_inference")
+	if ok {
+		role.AllowGCEInference = allowGCEInference.(bool)
+	} else if op == logical.CreateOperation {
+		role.AllowGCEInference = data.Get("allow_gce_inference").(bool)
+	}
+
+	maxJwtExp, ok := data.GetOk("max_jwt_exp")
+	if ok {
+		role.MaxJwtExp = time.Duration(maxJwtExp.(int)) * time.Second
+	} else {
```

Comment:
Did you mean this to be `else if op == logical.CreateOperation`? As is, anytime someone updates but doesn't specify this field it will reset the value.

Reply:
whoops, fixed